### ISSUE TITLE
Build fails when using gatsby-source-graphql in gatsby@next v4 beta.

### ISSUE THREAD

**Issue by wickedev** [gatsbyjs/gatsby#33302]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Build fails when using gatsby-source-graphql in gatsby@next v4 beta. The build was successful when I got a hint from the `warn Calling 'addThirdPartySchema' in the 'sourceNodes' API is deprecated. Please use: 'createSchemaCustomization'.` message and created `gatsby-node.js` in the project root, pasted the `gatsby-source-graphql` source code, and used `createSchemaCustomization` instead of `sourceNodes`.

Can I make a PR if the gatsby-source-graphql plugin can uses `createSchemaCustomization` instead of `sourceNodes` is ok?

### Reproduction Link

https://github.com/wickedev/bug-repro

### Steps to Reproduce

1. git clone https://github.com/wickedev/bug-repro.git
2. cd bug-repro
3. npm i
4. npm build

If you want to check the version using `createSchemaCustomization`, comment out the `gatsby-source-graphql` block in `gatsby-config.js` and `temp_gatsby-node.js` file rename to `gatsby-node.js` and then execute `npm run build`.

### Expected Result

```sh
C:\Users\wickedev\Workspace\bug-repro>npm run build

> gatsby-starter-hello-world@0.1.0 build
> gatsby build

success open and validate gatsby-configs, load plugins - 0.355s
success onPreInit - 0.002s

 ERROR 

(node:17944) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)

success delete worker cache from previous builds - 0.030s
success initialize cache - 0.154s
success copy gatsby files - 0.252s
success Compiling Gatsby Functions - 0.557s
success onPreBootstrap - 0.577s
success createSchemaCustomization - 0.130s
success Checking for changed pages - 0.003s
success source and transform nodes - 0.065s
info Writing GraphQL type definitions to C:/Users/wickedev/Workspace/bug-repro/.cache/schema.gql
success building schema - 0.758s
success createPages - 0.003s
success createPagesStatefully - 0.093s
info Total nodes: 21, SitePage nodes: 1 (use --verbose for breakdown)
success Checking for changed pages - 0.002s
success Cleaning up stale page-data - 0.005s
success update schema - 0.166s
success onPreExtractQueries - 0.002s
success extract queries from components - 0.370s
success write out redirect data - 0.004s
success onPostBootstrap - 0.002s
info bootstrap finished - 8.374s

 ERROR 

(node:22336) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)


 ERROR 

(node:16628) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)


 ERROR 

(node:24900) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)

success run queries in workers - 5.326s - 2/2 0.38/s
success Merge worker state - 0.004s
success write out requires - 0.023s
success Building production JavaScript and CSS bundles - 1.456s
success Caching JavaScript and CSS webpack compilation - 0.014s
success Rewriting compilation hashes - 0.007s
success Writing page-data.json files to public directory - 0.022s - 1/1 45.81/s
success Caching HTML renderer compilation - 0.003s
success Building HTML renderer - 0.586s
success Building static HTML for pages - 3.927s - 1/1 0.25/s
success onPostBuild - 0.002s
info Done building in 20.0321638 sec
```

### Actual Result

```sh
C:\Users\wickedev\Workspace\bug-repro>npm run build

> gatsby-starter-hello-world@0.1.0 build
> gatsby build

success open and validate gatsby-configs, load plugins - 0.364s
success onPreInit - 0.003s

 ERROR 

(node:18248) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)

success delete worker cache from previous builds - 0.031s
success initialize cache - 0.188s
success copy gatsby files - 0.271s
success Compiling Gatsby Functions - 0.551s
success onPreBootstrap - 0.574s
success createSchemaCustomization - 0.002s
warn Calling `addThirdPartySchema` in the `sourceNodes` API is deprecated. Please use: `createSchemaCustomization`.
success Checking for changed pages - 0.002s
success source and transform nodes - 0.226s
info Writing GraphQL type definitions to C:/Users/wickedev/Workspace/bug-repro/.cache/schema.gql
success building schema - 0.758s
success createPages - 0.005s
success createPagesStatefully - 0.100s
info Total nodes: 23, SitePage nodes: 1 (use --verbose for breakdown)
success Checking for changed pages - 0.002s
success Cleaning up stale page-data - 0.016s
success update schema - 0.155s
success onPreExtractQueries - 0.002s
success extract queries from components - 0.350s
success write out redirect data - 0.003s
success onPostBootstrap - 0.002s
info bootstrap finished - 8.354s

 ERROR 

(node:23512) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)


 ERROR 

(node:22756) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)


 ERROR 

(node:21144) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./public/" in the "exports" field module resolution of the package at C:\Users\wickedev\Workspace\bug-repro\node_modules\extract-files\package.json.Update this package.json to use a subpath pattern like "./public/*".
(Use `node --trace-deprecation ...` to show where the warning was created)


 ERROR #85923  GRAPHQL

There was an error in your GraphQL query:

Cannot query field "velog" on type "Query".

If you don't expect "velog" to exist on the type "Query" it is most likely a typo.
However, if you expect "velog" to exist there are a couple of solutions to common problems:

- If you added a new data source and/or changed something inside gatsby-node.js/gatsby-config.js, please try a restart of your development server
- The field might be accessible in another subfield, please try your query in GraphiQL and use the GraphiQL explorer to see which fields you can query and what shape they have
- You want to optionally use your field "velog" and right now it is not used anywhere. Therefore Gatsby can't infer the type and add it to the GraphQL schema. A quick fix is to add at least one entry with that field ("dummy  
content")

It is recommended to explicitly type your GraphQL schema if you want to use optional fields. This way you don't have to add the mentioned "dummy content". Visit our docs to learn how you can define the schema for "Query":    
https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization#creating-type-definitions


 ERROR #85928 

An error occurred during parallel query running.
Go here for troubleshooting tips: https://gatsby.dev/pqr-feedback



  Error: Worker exited before finishing task

  - index.js:112 ChildProcess.<anonymous>
    [bug-repro]/[gatsby-worker]/dist/index.js:112:45

  - node:events:394 ChildProcess.emit
    node:events:394:28

  - child_process:290 Process.ChildProcess._handle.onexit
    node:internal/child_process:290:12


not finished run queries in workers - 4.919s
```

### Environment

```shell
System:
    OS: Windows 10 10.0.19042
    CPU: (8) x64 Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
  Binaries:
    Node: 16.5.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 7.19.1 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.9.6 - /c/Python39/python
  Browsers:
    Chrome: 93.0.4577.82
    Edge: Spartan (44.19041.1023.0), Chromium (93.0.961.52)
  npmPackages:
    gatsby: ^4.0.0-zz-next.1 => 4.0.0-zz-next.1
    gatsby-source-graphql: ^4.0.0-zz-next.1 => 4.0.0-zz-next.8+fa70491bb6
  npmGlobalPackages:
    gatsby-cli: 4.0.0-zz-next.1
```

### Config Flags

```js
const fs = require('fs')
const { buildClientSchema } = require('graphql')

/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.com/docs/gatsby-config/
 */

module.exports = {
  /* Your site config here */
  plugins: [
    {
      resolve: "gatsby-source-graphql",
      options: {
        typeName: "Velog",
        fieldName: "velog",
        url: "https://v2.velog.io/graphql",
        createSchema: async () => {
          const json = JSON.parse(
            fs.readFileSync(`external-schema/velog-introspection.json`)
          )
          return buildClientSchema(json)
        },
      }
    },
  ],
}
```

**Comment by vladar** [comment]
Thanks for reporting. This should be fixed via https://github.com/gatsbyjs/gatsby/pull/33310

**Comment by laurenskling** [comment]
While giving Gatsby 4 a spin using `gatsby-source-graphql`, I found that DSG is not working :(. I've set my codebase https://github.com/laurenskling/using-gatsby-source-graphql to defer rendering the film pages, and now these pages get a `errors` prop. 

Example on Gatsby Cloud: https://build-58701a40-9907-423a-9f26-9f80e4b3beea.gtsb.io/ZmlsbXM6Mg== Saying: `Cannot query field "swapi" on type "Query".` So the entire external Graphql isn't found.

Building and serving locally I do no get this specific error, but another one. I'm getting:
```
{
  message: "nodeFetch is not a function"
  path: (2) ['swapi', 'film']
}
``` 
in my `errors` prop.
